ui/src/testing/stubs/cohorts-api-stub.ts

```diff
@@ -0,0 +1,142 @@
+import {convertToResources, ResourceType} from 'app/utils/resourceActionsReact';
+import {
+  Cohort,
+  CohortsApi,
+  EmptyResponse,
+  RecentResource,
+  Workspace,
+  WorkspaceAccessLevel
+} from 'generated/fetch';
+import {CohortListResponse} from 'generated/fetch/api';
+import {WorkspaceStubVariables} from './workspace-service-stub';
+
+export let DEFAULT_COHORT_ID = 1;
+export let DEFAULT_COHORT_ID_2 = 2;
+
+class CohortStub implements Cohort {
+  id?: number;
+
+  name: string;
+
+  criteria: string;
+
+  type: string;
+
+  description?: string;
+
+  creator?: string;
+
+  creationTime?: number;
+
+  lastModifiedTime?: number;
+
+  workspaceId: string;
+
+  constructor(cohort: Cohort, wsid: string) {
+    this.creationTime = cohort.creationTime;
+    this.creator = cohort.creator;
+    this.criteria = cohort.criteria;
+    this.description = cohort.description;
+    this.id = cohort.id;
+    this.lastModifiedTime = cohort.lastModifiedTime;
+    this.name = cohort.name;
+    this.type = cohort.type;
+    this.workspaceId = wsid;
+  }
+}
+
+export class CohortsApiStub extends CohortsApi {
+  public workspaces: Workspace[];
+  public cohorts: CohortStub[];
+  public resourceList: RecentResource[];
+
+  constructor() {
+    super(undefined, undefined, (..._: any[]) => { throw Error('cannot fetch in tests'); });
+
+    const stubWorkspace: Workspace = {
+      name: WorkspaceStubVariables.DEFAULT_WORKSPACE_NAME,
+      id: WorkspaceStubVariables.DEFAULT_WORKSPACE_ID,
+      namespace: WorkspaceStubVariables.DEFAULT_WORKSPACE_NS
+    };
+
+    const exampleCohort: CohortStub = {
+      id: DEFAULT_COHORT_ID,
+      name: 'sample name',
+      description: 'sample description',
+      criteria: '',
+      type: '',
+      workspaceId: WorkspaceStubVariables.DEFAULT_WORKSPACE_ID,
+      creationTime: new Date().getTime(),
+      lastModifiedTime: new Date().getTime() - 1000,
+    };
+
+    const exampleCohortTwo: CohortStub = {
+      id: DEFAULT_COHORT_ID_2,
+      name: 'sample name 2',
+      description: 'sample description 2',
+      criteria: '',
+      type: '',
+      workspaceId: WorkspaceStubVariables.DEFAULT_WORKSPACE_ID,
+      creationTime: new Date().getTime(),
+      lastModifiedTime: new Date().getTime() - 4000,
+    };
+    this.cohorts = [exampleCohort, exampleCohortTwo];
+    this.workspaces = [stubWorkspace];
+    this.resourceList = convertToResources(this.cohorts, stubWorkspace.namespace,
+      stubWorkspace.id, WorkspaceAccessLevel.OWNER, ResourceType.COHORT);
+  }
+
+  updateCohort(ns: string, wsid: string, cid: number, newCohort: Cohort): Promise<Cohort> {
+    return new Promise<Cohort>((resolve, reject) => {
+      setTimeout(() => {
```

Comment:
You can remove all of these `setTimeout()`s in the stub now. This was a workaround to Observable's inconsistent behavior around callback resolution. Namely if a value is present on an observable, a callbacks are executed synchronously. So without the setTimeout, the following would have worked in tests, but failed in a real browser:

```
let foo;
cohortsApi.updateCohort.subscribe(resp => {
  foo = resp;
});
console.log(foo.items);
```

Promises always execute callbacks asynchronously, so it's not a problem here - e.g. try this in Chrome console:

```
(() => {
    Promise.resolve(1).then(console.log);;
    console.log(2)
})();
2
1
```

Reply:
👍